gpu/containers/include/pcl/gpu/containers/impl/device_array.hpp

```diff
@@ -65,7 +65,20 @@ template<class T> inline void DeviceArray<T>::upload(const T *host_ptr, std::siz
 { DeviceMemory::upload(host_ptr, size * elem_size); }
 template<class T> inline void DeviceArray<T>::download(T *host_ptr) const
 { DeviceMemory::download( host_ptr ); }
-
+template <class T>
+inline bool
+DeviceArray<T>::download(T* host_ptr,
+                         std::size_t device_begin_offset,
+                         std::size_t device_end_offset) const
+{
+  if (device_end_offset < device_begin_offset) {
+    return false;
+  }
+  const T* begin = ptr() + device_begin_offset;
```

Comment:
Can we move this to DeviceMemory as well? That'll keep DeviceMemory from downloading from a random location accidentally

Reply:
Thanks for your comment! 

I think this would be possible by altering the function signature of `DeviceMemory::download`. The `DeviceMemory` class acts as blob storage and does not know the type of elements it stores. Defining memory locations to sync within member functions is thus not feasible currently. However, if the API of the memory device was `DeviceMemory::Download(host_ptr, device_begin_offset, device_end_offset, elem_size)`, we can include the lines you mentioned in the DeviceMemory class. However, I am not sure this is useful. Firstly, our cudaSafeCall macros wrapping Cuda interactions report errors, and users can guard themselves against this by using try-catch statements. Secondly, I did not see any PCL code using DeviceMemory directly. Instead, all interactions are through the DeviceArray class. What do you think about that?

Reply:
> users can guard themselves against this by using try-catch statements.

If there's a potentially safer API where such counter-measures aren't required on the user's end, then we should aim for that. PCL has a lot of baggage, let's not add another item for a future cleanup 😆 

> Secondly, I did not see any PCL code using DeviceMemory directly

It's public API. Even if we don't use it, someone might be using it. Another point is that if we can have a consistent API, then it's friendlier for downstream users.

Since `DeviceMemory` (DM) works on bytes, and not datatype `T`, we can have the following API to mirror the `DeviceArray` (DA) API:
```cpp
DM::download(host_ptr, device_begin_byte_offset, device_end_byte_offset)
```

My comments are relevant mostly to the public API. The private/protected API can be more raw. Eg:
```erl
DA::download/3: check_offset/2 && DM::download_protected/3
DM::download/3: check_offset/2 && DM::download_protected/3
DM::download_protected_3: cudamemcpy/4 && cudaSync0
```
(please forgive the weird syntax used)

Reply:
Thanks for your reply! Your suggestion for the `DM::download` signature is very useful and I will try to implement it - thank you!

Reply:
Thanks again for your comments, @kunaltyagi! I tried to address your concerns, but I am not sure my implementation is proper. As the device memory is type agnostic, we cannot perform pointer arithmetic to calculate the position of the data to be downloaded. Instead, I cast the `void*` pointer to a `char*` to perform the required arithmetic. I am not sure if this is idiomatic code or if a better solution exists. What do you think about the implementation?

Reply:
The cast to char* ann back to void* is inevitable due to the existing code. Do you think we need to make the conversion to void* explicit (using static_cast)?

Reply:
Fantastic - thank you so much for your review, @kunaltyagi !